store/iterator.go

```diff
@@ -108,163 +95,158 @@ func (b *btreeIter) wrap(parent Iterator) *itemIter {
+		return nil, errors.Wrap(errors.ErrDone, "btree iterator")
+	}
+	key, ok := data.(keyer)
+	if !ok {
+		return nil, errors.Wrapf(errors.ErrType, "expected keyer, got %T", data)
+	}
+	return key, nil
 }
 
-func (b *btreeIter) close() {
+func (b *btreeIter) Release() {
 	b.once.Do(func() {
 		b.stop <- struct{}{}
 	})
 }
 
-// get requires this is valid, gets what we are pointing at
-func (b *btreeIter) get() keyer {
-	return b.data.(keyer)
-}
-
-func (b *btreeIter) valid() bool {
-	return b.hasMore
-}
-
 type itemIter struct {
 	wrap *btreeIter
 	// if we are iterating in a cache-wrap (and who isn't),
 	// we need to combine this iterator with the parent
 	parent Iterator
+
+	parentDone   bool
+	cachedParent Model
+	wrapDone     bool
+	cachedWrap   keyer
 }
 
 //------- public facing interface ------
 var _ Iterator = (*itemIter)(nil)
 
-// Valid implements Iterator and returns true iff it can be read
-func (i *itemIter) Valid() bool {
-	return i.wrap.valid() || i.parentValid()
-}
-
-// Next moves the iterator to the next sequential key in the database, as
-// defined by order of iteration.
+// advanceParent will read next from parent iterators,
+// and set cached value as well as done flags.
 //
-// If Valid returns false, this method will panic.
-func (i *itemIter) Next() error {
-	// advance either us, parent, or both
-	switch i.firstKey() {
-	case us:
-		i.wrap.next()
-	case both:
-		i.wrap.next()
-		fallthrough
-	case parent:
-		err := i.parent.Next()
-		if err != nil {
-			return err
+// it will skip closed and missing iterators.
+// doesn't return ErrDone, but only unexpected data errors.
+func (i *itemIter) advanceParent() error {
+	if i.parent == nil {
+		i.parentDone = true
+	}
+
+	if !i.parentDone && i.cachedParent.Key == nil {
+		key, value, err := i.parent.Next()
+		if errors.ErrDone.Is(err) {
+			i.parentDone = true
+		} else if err != nil {
+			return errors.Wrap(err, "advance parent")
+		} else {
+			i.cachedParent = weave.Model{Key: key, Value: value}
 		}
-	default:
-		panic("Advanced past the end!")
 	}
 
-	// keep advancing over all deleted entries
-	return i.skipAllDeleted()
+	return nil
 }
 
-// Key returns the key of the cursor.
-func (i *itemIter) Key() (key []byte) {
-	switch i.firstKey() {
-	case us, both:
-		return i.wrap.get().Key()
-	case parent:
-		return i.parent.Key()
-	default: //none
-		panic("Advanced past the end!")
+// advance will read next from wrap iterators,
+// and set cached value as well as done flags.
+//
+// It will skip any deleted items before the i.cachedParent.Key value
+//
+// it will skip closed and missing iterators.
+// doesn't return ErrDone, but only unexpected data errors.
+func (i *itemIter) advanceWrap() error {
+	for !i.wrapDone && i.cachedWrap == nil {
+		var err error
+		i.cachedWrap, err = i.wrap.Next()
+		if errors.ErrDone.Is(err) {
+			i.wrapDone = true
+		} else if err != nil {
+			return errors.Wrap(err, "advance wrap")
+		} else if _, ok := i.cachedWrap.(deletedItem); ok {
```

Comment:
PP: if you `return nil` in L171 you can drop all the `else` and make it a bit more readable/ Go like